src/analysis/liveness-transfer-function.h

```diff
@@ -0,0 +1,101 @@
+#ifndef wasm_analysis_liveness_transfer_function_h
+#define wasm_analysis_liveness_transfer_function_h
+
+#include "lattice.h"
+#include "monotone-analyzer.h"
+
+namespace wasm::analysis {
+
+class LivenessTransferFunction : public Visitor<LivenessTransferFunction> {
+  FinitePowersetLattice::Element* currState;
+
+public:
+  LivenessTransferFunction() : currState(nullptr) {}
+
+  // Transfer function implementation. Modifies the state for a particular
+  // expression type. In our current limited implementation, we just update
+  // state on gets and sets of local indices.
+  void visitLocalSet(LocalSet* curr) {
+    if (currState) {
+      currState->set(curr->index, false);
+    }
+  }
+  void visitLocalGet(LocalGet* curr) {
+    if (currState) {
+      currState->set(curr->index, true);
+    }
+  }
+
+  // Executes the transfer function on all the expressions of the corresponding
+  // CFG node, starting with the node's input state. Returns the final output
+  // state of the node.
```

Comment:
This description is very generic, and indeed it looks like there is nothing specific to liveness in the function. Could this not be handled by the parent? That is, the parent could have a flag whether the analysis is forward or backwards (or we could have a ForwardAnalysis and BackwardAnalysis). Then all `LivenessTransferFunction` would contain is just the LocalGet/LocalSet logic, which is all that is specific to liveness.

Likewise, in this idea, `enqueueWorklist` can also be in the parent, which again I think makes sense since it just depends on the order of the analysis.

My motivation here is that we'll have many backwards analyses, and it would be good that we don't duplicate the backwards-ordering code in them all.

Reply:
I agree that there is a lot of generic forward/backward functionality which could be re-used. My reasons for doing this are:

- It abstracts away the visitor implementation, in case in future transfer functions, we want to use a different implementation, keep state between expressions, or select expressions types differently (e.g. through casts).
- Right now, the state is set as a pointer member variable of the transfer function so that it can be accessed by the visitor functions without needing it to be passed as a parameter and thus changing the visitor API. It seems cleaner to set it once, mutate it a bunch of times, and then be done instead of resetting the pointer every time we visit an expression.
- The current implementation means nothing has to be parameterized in the parent, and the user is free to implement any arbitrary custom analysis flow, not just forward or backward. Admittedly, there are drawbacks to this, as you pointed out.
- The CFG basic block level seemed a good delineation point between the transfer function and the parent.

I'm not entirely decided about this method either.

Reply:
It does seem like a good idea to have a `VisitorTransferFunction` CRTP utility that would handle everything here except for the `visit` methods themselves. I imagine that `VisitorTransferFunction` could have a template parameter saying whether it is a forward or backward analysis (or neither) and that it could provide a default implementation of `depsBegin` and `depsEnd` accordingly, which the particular subclass would still be able to replace if necessary.

Reply:
Saved for later PR.